python/tk_multi_breakdown/breakdown.py

```diff
@@ -0,0 +1,240 @@
+                # method 'register_publish'
+                for key_name, key in matching_template.keys.iteritems():
+                    if key_name in fields and key.is_abstract:
+                        del(fields[key_name])
+
+                # we also want to normalize the eye field (this should probably be an abstract field!)
+                # note: we need to do this explicitly because the eye isn't abstract in the default 
+                # configs yet (which is incorrect!).
+                fields["eye"] = "%V"
+                
+                # now build the normalized path that we can use to find corresponding Shotgun published files
+                normalized_path = matching_template.apply_fields(fields)
+                
+                item = {}
+                item["path"] = normalized_path
+                item["node_name"] = node_name
+                item["node_type"] = node_type
+                item["template"] = matching_template
+                item["fields"] = fields
+                item["sg_data"] = None
+
+
+                # store the normalized fields in dict
+                items.append(item)
+
+    # now do a second pass on all the files that are valid to see if they are published
+    # note that we store (by convention) all things on a normalized sequence form in SG, e.g
+    # all four-padded sequences are stored as '%04d' regardless if they have been published from
+    # houdini, maya, nuke etc.
+    valid_paths = [ x.get("path") for x in items ]
+
+    # check if we have the path in the cache
+    paths_to_fetch = []
+    for p in valid_paths:
+        if p not in g_cached_sg_publish_data:
+            paths_to_fetch.append(p)
+        else:
+            # use cache data!
+            for item in items:
+                if item.get("path") == p:
+                    item["sg_data"] = g_cached_sg_publish_data[p]
+
+
+    fields = ["entity",
+              "entity.Asset.sg_asset_type", # grab asset type if it is an asset
+              "code",
+              "image",
+              "name",
+              "task",
+              "version_number",
+              "project"
+              ]
+
+    if tank.util.get_published_file_entity_type(app.tank) == "PublishedFile":
+        fields.append("published_file_type")
+    else:# == "TankPublishedFile"
+        fields.append("tank_type")
+
+    sg_data = tank.util.find_publish(app.tank, paths_to_fetch, fields=fields)
+
+    # process and cache shotgun items
+    for (path, sg_chunk) in sg_data.items():
+        # cache item
+        g_cached_sg_publish_data[path] = sg_chunk
+
+        # append the sg data to the right path
+        for item in items:
+            if item.get("path") == path:
+                item["sg_data"] = sg_chunk
+
+    # we no longer need the path key in the dict, so get rid of it
+    for item in items:
+        del item["path"]
+
+    return items
+
+
+def compute_highest_version(template, curr_fields):
```

Comment:
Given a scene with hundreds or thousands of reference to the same file (a crowd for example or a tree in a forest), you will be scanning files on disk forever, especially if there are multiple versions for that file. Seems like some caching is in order.